# Comparing RBP Analyses from Ferrarini et al. and [Schmidt et al.](https://www.nature.com/articles/s41564-020-00846-z)

In [1]:
!pip install biopython
!pip install pyensembl

     |████████████████████████████████| 2.3 MB 4.5 MB/s 
     |████████████████████████████████| 55 kB 2.2 MB/s 
     |████████████████████████████████| 130 kB 8.5 MB/s 
  Created wheel for pyensembl: filename=pyensembl-1.9.4-py3-none-any.whl size=56443 sha256=dc2515e8af0f8199a83fb60ae5a4051702c498b076f0e9bc932064bf1e0bf360
  Stored in directory: /root/.cache/pip/wheels/14/e7/fe/08360ab53ff9bebb6daf956827c04e894d76dc0f64eb7502ef
  Created wheel for datacache: filename=datacache-1.1.5-py3-none-any.whl size=16473 sha256=3e34a829b5ea4ea81f562c05fedb7b3ac454e1ca91d9d1906d96830bc1bbe94e
  Stored in directory: /root/.cache/pip/wheels/83/dd/38/d52c1ed3ad3ec655ea30a429de1cf89fac1009e8cb0365703a
  Created wheel for gtfparse: filename=gtfparse-1.2.1-py3-none-any.whl size=11637 sha256=6d9bd04f006175dacac80f42711a6be6ba7b6faa7a60d0c2338ad6bfffe78be6
  Stored in directory: /root/.cache/pip/wheels/a9/7a/45/434ba7d04afbfcc5c12195682413f53ecd4580dfd62905a672
  Created wheel for memoized-property: file

In [2]:
from google.colab import drive

drive.mount("/content/drive")

Mounted at /content/drive


In [3]:
from pathlib import Path

CONF_DATA = Path("/content/drive/MyDrive/ecbm-e4060-covid-interactions/DATA/Hackathon/output/ki-results")
LIT_DATA = Path("/content/drive/MyDrive/ecbm-e4060-covid-interactions/DATA/Literature")
HACK_DATA = Path("/content/drive/MyDrive/ecbm-e4060-covid-interactions/DATA/Hackathon")

## Comparing Our Output to Schmidt et al.

In [4]:
genome_hits_file = "genome_hits.tsv"
f_utr_hits_file = "f_utr_hits.tsv"
t_utr_hits_file = "t_utr_hits.tsv"
genome_neg_hits_file = "genome_neg_hits.tsv"

In [5]:
import pandas as pd

genome_hits = pd.read_csv(CONF_DATA / genome_hits_file, sep="\t")
f_utr_hits = pd.read_csv(CONF_DATA / f_utr_hits_file, sep="\t")
t_utr_hits = pd.read_csv(CONF_DATA / t_utr_hits_file, sep="\t")
genome_neg_hits = pd.read_csv(CONF_DATA / genome_neg_hits_file, sep="\t")

In [6]:
total_hits = pd.concat([genome_hits, f_utr_hits, t_utr_hits, genome_neg_hits])
total_hits.Gene_name.unique()

array(['HNRNPL', 'FUS', 'MBNL1', 'SRSF1', 'RBMY1A1', 'ZFP36', 'SRSF10',
       'PTBP1', 'YBX2', 'SRSF3', 'PABPC1', 'PABPN1', 'SART3', 'PABPC4',
       'ZNF638', 'PABPC5', 'PABPC3', 'CELF2', 'YBX1', 'ZRANB2', 'KHDRBS3',
       'LIN28A', 'PPIE', 'HNRNPA1', 'NOVA2', 'NOVA1', 'CELF6', 'ELAVL1',
       'TIA1', 'CELF4', 'HNRNPDL', 'RBFOX1'], dtype=object)

In [7]:
total_hits

Gene_name strand     N  ...          z           pval           qval
1     HNRNPL      +   632  ...  12.075008   7.157531e-34   7.085956e-32
2        FUS      +   140  ...  10.552335   2.477497e-26   1.226361e-24
3      MBNL1      +   682  ...   9.353181   4.252541e-21   1.403339e-19
4      SRSF1      +   335  ...   7.488976   3.470646e-14   8.589849e-13
5    RBMY1A1      +   107  ...   6.987404   1.400092e-12   2.772182e-11
6      ZFP36      +   609  ...   6.782399   5.909821e-12   9.751205e-11
7     SRSF10      +    88  ...   6.422941   6.683299e-11   9.452095e-10
8      PTBP1      +  3151  ...   5.920488   1.604938e-09   1.986111e-08
9       YBX2      +    51  ...   5.453245   2.472947e-08   2.720242e-07
10     SRSF3      +    74  ...   5.425103   2.896061e-08   2.867100e-07
11    PABPC1      +   118  ...   4.929721   4.117351e-07   3.705616e-06
12    PABPN1      +    50  ...   4.818234   7.241708e-07   5.974409e-06
13     SART3      +    49  ...   4.536058   2.865765e-06   2.182390e-05
14    PABPC4      +    28  ...   3.935445   4.152128e-05   2.936148e-04
15    ZNF638      +    27  ...   3.908469   4.644150e-05   3.065139e-04
16    PABPC5      +    33  ...   3.685893   1.139509e-04   7.050710e-04
17    PABPC3      +    34  ...   3.431014   3.006647e-04   1.750930e-03
18     CELF2      +   698  ...   3.410527   3.241872e-04   1.783030e-03
19      YBX1      +   204  ...   3.160143   7.884578e-04   4.108280e-03
1     ZRANB2      +     4  ...   3.342141   4.156747e-04   3.948909e-03
1     PABPC4      +    27  ...  25.108129  2.026962e-139  7.499760e-138
2      SART3      +    28  ...  22.594349  2.462593e-113  4.555796e-112
3     PABPC1      +    28  ...  18.911272   4.605438e-80   5.680040e-79
4     SRSF10      +    27  ...  18.695139   2.711437e-78   2.508079e-77
5    KHDRBS3      +    28  ...  14.517585   4.687752e-48   3.468937e-47
6     LIN28A      +     3  ...   6.043128   7.557728e-10   4.660599e-09
7       PPIE      +    42  ...   5.088977   1.800000e-07   7.399998e-07
8    HNRNPA1      +     4  ...   3.740120   9.196624e-05   3.402751e-04
1      NOVA2      -    95  ...  12.370490   1.887545e-35   1.812043e-33
2       YBX1      -   361  ...  11.486870   7.678840e-31   3.685843e-29
3      NOVA1      -   918  ...  11.228229   1.481794e-29   4.741741e-28
4     HNRNPL      -   769  ...  11.058703   9.947639e-29   2.387433e-27
5       YBX2      -    92  ...   9.917461   1.747669e-23   3.355524e-22
6      CELF2      -   619  ...   6.988200   1.392173e-12   2.227477e-11
7      SRSF3      -    88  ...   6.241641   2.165026e-10   2.969178e-09
8      CELF6      -    30  ...   5.069709   1.992126e-07   2.390551e-06
9     ELAVL1      -   104  ...   4.157687   1.607432e-05   1.714594e-04
10      TIA1      -   998  ...   4.110909   1.970524e-05   1.891703e-04
11     CELF4      -    20  ...   3.776978   7.937142e-05   6.719824e-04
12       FUS      -    72  ...   3.762839   8.399780e-05   6.719824e-04
13     ZFP36      -   505  ...   3.226549   6.264636e-04   4.626193e-03
14   HNRNPDL      -   209  ...   3.140942   8.420275e-04   5.773903e-03
15    RBFOX1      -    51  ...   3.003744   1.333400e-03   8.533760e-03

[43 rows x 8 columns]

In [8]:
#The SARS-CoV-2 RNA–protein interactome in infected human cells, December 2020
#supplementary table 1 (Proteins detected by quantitative mass spectrometry in 
#SARS-CoV-2 RNA and RMRP RNA antisense purifications in infected human cells.)

st1 = LIT_DATA / "41564_2020_846_MOESM2_ESM-1.xlsx"
st1_pd = pd.read_excel(st1, index_col=0, skiprows=2)
#st1_pd[(st1_pd['adj.P.Val.SCoV2.over.RMRP'] < 0.05) & (st1_pd['logFC.SCoV2.over.RMRP'] > 0)]

In [9]:
total_hits[total_hits.Gene_name.isin(st1_pd.geneSymbol)].Gene_name.nunique()

9

In [10]:
conf_schmidt = total_hits[total_hits.Gene_name.isin(st1_pd.geneSymbol)]

## Comparing Ferrarini et al. Output to Schmidt et al.

Before performing this analysis, I placed the significantly enriched RBPs from Ferrarini et al.'s paper (listed on pages 10 and 11) in TSV files.

In [11]:
ferrarini_data = pd.read_csv(HACK_DATA / "enriched_rbp_list.tsv", sep="\t", index_col=0)

In [12]:
ferrarini_data[ferrarini_data.Gene_name.isin(st1_pd.geneSymbol)].Gene_name.nunique()

10

In [13]:
ferrarini_schmidt = ferrarini_data[ferrarini_data.Gene_name.isin(st1_pd.geneSymbol)]

In [14]:
set(ferrarini_schmidt.Gene_name) - set(conf_schmidt.Gene_name)

{'HNRNPA2B1'}

In [15]:
ferrarini_schmidt

Gene_name Binding_site_location
0      HNRNPL                Genome
3       SRSF1                Genome
7       PTBP1                Genome
8       SRSF3                Genome
10     PABPC1                Genome
13     PABPC4                Genome
18       YBX1                Genome
23     PABPC4                 3'UTR
25     PABPC1                 3'UTR
29  HNRNPA2B1                 3'UTR
32    HNRNPA1                 3'UTR
35       YBX1       Negative Strand
36     HNRNPL       Negative Strand
40      SRSF3       Negative Strand
47    HNRNPDL       Negative Strand

## Grabbing Protein Sequences from Schmidt et al. Data

In [16]:
relaxed_fdr = st1_pd[(st1_pd['adj.P.Val.SCoV2.over.RMRP'] < 0.2) & (st1_pd['logFC.SCoV2.over.RMRP'] > 0)]

This above criteria were used by Schmidt et al. to identify the 104 RBPs they say bind to the SARS-CoV-2 genome. `relaxed_fdr` contains 12 extra RBPs that come from the SARS-CoV-2 species instead of Homo Sapiens. The `accession_number` field can be used to obtain their sequences using Entrez.

In [17]:
relaxed_fdr

adj.P.Val.SCoV2.over.RMRP  ...                                         entry_name
id                                         ...                                                   
NP_001120665.1                   0.000111  ...   cellular nucleic acid-binding protein isoform...
YP_009742613.1                   0.000290  ...   mat_peptide:nsp6:ORF1a_polyprotein:ORF1ab:GN=...
NP_055935.4                      0.136832  ...                Alstrom syndrome protein 1 GN=ALMS1
NP_001129125.1                   0.002919  ...   polyadenylate-binding protein 4 isoform 1 GN=...
YP_009725310.1                   0.002303  ...   mat_peptide:endoRNAse:ORF1b_polyprotein:ORF1a...
...                                   ...  ...                                                ...
NP_001005335.1                   0.156721  ...   heterogeneous nuclear ribonucleoprotein L iso...
NP_001092104.1                   0.179775  ...          RNA-binding protein 47 isoform a GN=RBM47
NP_001129171.1                   0.183259  ...                 14-3-3 protein zeta/delta GN=YWHAZ
NP_000166.2                      0.181670  ...     glucose-6-phosphate isomerase isoform 2 GN=GPI
NP_000963.1                      0.186766  ...                 60S ribosomal protein L7a GN=RPL7A

[118 rows x 22 columns]

### Positive Examples

In [18]:
relaxed_fdr.accession_number.unique()

array(['NP_001120665.1', 'YP_009742613.1', 'NP_055935.4',
       'NP_001129125.1', 'YP_009725310.1', 'NP_001138898.1',
       'YP_009724397.2', 'NP_003746.2', 'YP_009724390.1',
       'NP_001166299.1', 'NP_001164274.1', 'NP_005889.3',
       'YP_009724391.1', 'NP_003747.1', 'NP_001004317.1',
       'NP_001269581.1', 'NP_004550.2', 'NP_001018494.1',
       'NP_001185747.1', 'NP_001071105.1', 'NP_150093.1', 'NP_055121.1',
       'NP_036429.2', 'NP_001153146.1', 'YP_009725307.1', 'NP_002559.2',
       'YP_009725311.1', 'YP_009724393.1', 'NP_055205.2',
       'NP_001284478.1', 'NP_001287750.1', 'YP_009742616.1',
       'NP_001230648.1', 'NP_001231903.1', 'NP_001007554.1',
       'NP_001123551.1', 'NP_000961.2', 'NP_071496.1', 'NP_002943.2',
       'NP_002889.1', 'NP_001005.1', 'NP_001180345.1', 'NP_057190.2',
       'NP_001020241.1', 'NP_001005386.1', 'NP_005498.1', 'NP_001559.1',
       'NP_009109.3', 'NP_001107565.1', 'NP_002127.1', 'NP_001295206.1',
       'NP_001276978.1', 'NP_00100722

The following is a list of RBPs from [Sundararaman et al.](https://www.sciencedirect.com/science/article/pii/S1097276516000964?via%3Dihub).

In [19]:
rbp_list_file = LIT_DATA / "1-s2.0-S1097276516000964-mmc2.xlsx"
rbp_list = pd.read_excel(rbp_list_file, sheet_name="Sheet2_1072_RBP_compilation")
rbp_list.columns = ["ensembl_ID", "gene_symbol", "gene_description"]
rbp_list

ensembl_ID  ...                                   gene_description
0     ENSG00000090861  ...                             alanyl-tRNA synthetase
1     ENSG00000275700  ...        apoptosis antagonizing transcription factor
2     ENSG00000204574  ...  ATP-binding cassette, sub-family F (GCN20), me...
3     ENSG00000146109  ...                 activator of basal transcription 1
4     ENSG00000167315  ...                       acetyl-CoA acyltransferase 2
...               ...  ...                                                ...
1067  ENSG00000124201  ...                zinc finger, NFX1-type containing 1
1068  ENSG00000132485  ...       zinc finger, RAN-binding domain containing 2
1069  ENSG00000212643  ...  zinc finger (CCCH type), RNA-binding motif and...
1070  ENSG00000169249  ...  zinc finger (CCCH type), RNA-binding motif and...
1071  ENSG00000159840  ...                                              zyxin

[1072 rows x 3 columns]

I subset for those not identified by Schmidt et al. I'll randomly select 118 of these to be treated as the negative set.

In [20]:
negative_total = rbp_list[~rbp_list.gene_symbol.isin(st1_pd.geneSymbol)]
negative_total

ensembl_ID  ...                                   gene_description
0     ENSG00000090861  ...                             alanyl-tRNA synthetase
1     ENSG00000275700  ...        apoptosis antagonizing transcription factor
2     ENSG00000204574  ...  ATP-binding cassette, sub-family F (GCN20), me...
3     ENSG00000146109  ...                 activator of basal transcription 1
7     ENSG00000164113  ...  adenosine deaminase domain containing 1 (testi...
...               ...  ...                                                ...
1067  ENSG00000124201  ...                zinc finger, NFX1-type containing 1
1068  ENSG00000132485  ...       zinc finger, RAN-binding domain containing 2
1069  ENSG00000212643  ...  zinc finger (CCCH type), RNA-binding motif and...
1070  ENSG00000169249  ...  zinc finger (CCCH type), RNA-binding motif and...
1071  ENSG00000159840  ...                                              zyxin

[860 rows x 3 columns]

In [21]:
sampled = negative_total.sample(n=118)
sampled

ensembl_ID  ...                                   gene_description
475  ENSG00000086504  ...                mitochondrial ribosomal protein L28
705  ENSG00000265241  ...                       RNA binding motif protein 8A
531  ENSG00000130935  ...                               nucleolar protein 11
685  ENSG00000119707  ...                       RNA binding motif protein 25
50   ENSG00000224470  ...                                      ataxin 1-like
..               ...  ...                                                ...
466  ENSG00000133606  ...                      makorin ring finger protein 1
75   ENSG00000162642  ...                 chromosome 1 open reading frame 52
708  ENSG00000144642  ...  RNA binding motif, single stranded interacting...
822  ENSG00000156304  ...                 SR-related CTD-associated factor 4
260  ENSG00000120658  ...               ecto-NOX disulfide-thiol exchanger 1

[118 rows x 3 columns]

Need accession numbers for these...

In [22]:
list(sampled.ensembl_ID)

['ENSG00000086504',
 'ENSG00000265241',
 'ENSG00000130935',
 'ENSG00000119707',
 'ENSG00000224470',
 'ENSG00000113460',
 'ENSG00000033030',
 'ENSG00000104413',
 'ENSG00000055483',
 'ENSG00000130985',
 'ENSG00000080802',
 'ENSG00000114416',
 'ENSG00000151923',
 'ENSG00000055044',
 'ENSG00000149016',
 'ENSG00000100764',
 'ENSG00000105793',
 'ENSG00000180098',
 'ENSG00000118197',
 'ENSG00000131503',
 'ENSG00000198242',
 'ENSG00000132773',
 'ENSG00000164548',
 'ENSG00000148840',
 'ENSG00000205323',
 'ENSG00000119285',
 'ENSG00000163239',
 'ENSG00000139675',
 'ENSG00000170748',
 'ENSG00000129317',
 'ENSG00000075292',
 'ENSG00000118246',
 'ENSG00000197181',
 'ENSG00000106638',
 'ENSG00000113742',
 'ENSG00000137876',
 'ENSG00000134987',
 'ENSG00000166197',
 'ENSG00000126749',
 'ENSG00000132603',
 'ENSG00000168003',
 'ENSG00000086758',
 'ENSG00000183684',
 'ENSG00000138709',
 'ENSG00000003756',
 'ENSG00000134627',
 'ENSG00000188613',
 'ENSG00000149658',
 'ENSG00000171056',
 'ENSG00000108298',


In [25]:
# This code came from UniProt docs
import urllib.parse
import urllib.request

url = 'https://www.uniprot.org/uploadlists/'

params = {
'from': 'ENSEMBL_ID',
'to': 'SWISSPROT',
'format': 'tab',
'query': " ".join(list(sampled.ensembl_ID))
}

data = urllib.parse.urlencode(params)
data = data.encode('utf-8')
req = urllib.request.Request(url, data)
with urllib.request.urlopen(req) as f:
   response = f.read()
response = response.decode('utf-8')
print(response)

From	To
ENSG00000086504	Q13084
ENSG00000265241	Q9Y5S9
ENSG00000130935	Q9H8H0
ENSG00000119707	P49756
ENSG00000224470	P0C7T5
ENSG00000113460	Q8TDN6
ENSG00000033030	Q6NZY4
ENSG00000104413	Q6NXG1
ENSG00000055483	Q9P275
ENSG00000130985	P22314
ENSG00000080802	O95628
ENSG00000114416	P51114
ENSG00000151923	Q01085
ENSG00000055044	Q9Y2X3
ENSG00000149016	Q9H6E5
ENSG00000100764	P62191
ENSG00000105793	A4D1E9
ENSG00000180098	Q9NX07
ENSG00000118197	Q5T1V6
ENSG00000131503	Q8IWZ3
ENSG00000198242	P62750
ENSG00000132773	Q96GM8
ENSG00000164548	Q13595
ENSG00000148840	Q5VV67
ENSG00000205323	P82979
ENSG00000119285	Q9H583
ENSG00000163239	Q5VZ19
ENSG00000139675	Q32P51
ENSG00000170748	O75526
ENSG00000129317	Q9H0K6
ENSG00000075292	Q14966
ENSG00000118246	Q9NYY8
ENSG00000197181	Q8TC59
ENSG00000106638	Q9Y4P3
ENSG00000113742	Q17RY0
ENSG00000137876	Q9UHA3
ENSG00000134987	Q8NI36
ENSG00000166197	Q14978
ENSG00000126749	Q92979
ENSG00000132603	Q9Y221
ENSG00000168003	P08195
ENSG00000086758	Q7Z6Z7
ENSG00000138709	Q659C4
ENS

In [26]:
swiss_prot_ids = []
for line in response.strip().split("\n")[1:]:
    #print(line)
    sp_id = line.split("\t")[1]
    swiss_prot_ids.append(sp_id)
swiss_prot_ids

['Q13084',
 'Q9Y5S9',
 'Q9H8H0',
 'P49756',
 'P0C7T5',
 'Q8TDN6',
 'Q6NZY4',
 'Q6NXG1',
 'Q9P275',
 'P22314',
 'O95628',
 'P51114',
 'Q01085',
 'Q9Y2X3',
 'Q9H6E5',
 'P62191',
 'A4D1E9',
 'Q9NX07',
 'Q5T1V6',
 'Q8IWZ3',
 'P62750',
 'Q96GM8',
 'Q13595',
 'Q5VV67',
 'P82979',
 'Q9H583',
 'Q5VZ19',
 'Q32P51',
 'O75526',
 'Q9H0K6',
 'Q14966',
 'Q9NYY8',
 'Q8TC59',
 'Q9Y4P3',
 'Q17RY0',
 'Q9UHA3',
 'Q8NI36',
 'Q14978',
 'Q92979',
 'Q9Y221',
 'P08195',
 'Q7Z6Z7',
 'Q659C4',
 'P52756',
 'Q7Z3Z4',
 'Q8WY41',
 'Q9BYJ9',
 'Q9BT81',
 'P84098',
 'Q5U5Q3',
 'O75607',
 'O60869',
 'Q15477',
 'Q9HAV4',
 'Q96KR1',
 'Q7Z478',
 'Q8NDT2',
 'Q15428',
 'Q96M93',
 'P18124',
 'Q9NX74',
 'P62979',
 'P05388',
 'Q8N5A5',
 'Q9UBU9',
 'Q1ED39',
 'Q99848',
 'Q5VWX1',
 'Q86TM3',
 'P23246',
 'Q587J7',
 'Q9NQC3',
 'Q9BRU9',
 'O43399',
 'Q9NPE3',
 'Q15696',
 'Q9NV31',
 'Q9Y324',
 'P42285',
 'O76021',
 'O75525',
 'Q9BQ75',
 'Q8WVV9',
 'Q9H8H2',
 'Q14684',
 'Q9Y3U8',
 'Q13148',
 'Q8NBP7',
 'O60812',
 'O75475',
 'Q96T37',

In [27]:
import os
import sys

from urllib.request import urlretrieve

import Bio
from Bio import SeqIO, SearchIO, Entrez
from Bio.Seq import Seq
from Bio.SeqUtils import GC
from Bio.Blast import NCBIWWW
from Bio.Data import CodonTable

In [28]:
from Bio import Entrez
Entrez.email = "sg4040@columbia.edu"

In [29]:
#retrieve S protein sequences 
handle = Entrez.efetch(db="protein", id=list(relaxed_fdr.accession_number), rettype="fasta")

#save sequences to fasta file
fasta_file = "positive_proteins.fasta"
with open(fasta_file, "w") as fo:
    fo.write(handle.read())

#check that everything worked
!cat positive_proteins.fasta

>NP_001120665.1 CCHC-type zinc finger nucleic acid binding protein isoform 2 [Homo sapiens]
MSSNECFKCGRSGHWARECPTGGGRGRGMRSRGRGGFTSDRGFQFVSSSLPDICYRCGESGHLAKDCDLQ
EDEACYNCGRGGHIAKDCKEPKREREQCCYNCGKPGHLARDCDHADEQKCYSCGEFGHIQKDCTKVKCYR
CGETGHVAINCSKTSEVNCYRCGESGHLARECTIEATA

>YP_009742613.1 nsp6 [Severe acute respiratory syndrome coronavirus 2]
SAVKRTIKGTHHWLLLTILTSLLVLVQSTQWSLFFFLYENAFLPFAMGIIAMSAFAMMFVKHKHAFLCLF
LLPSLATVAYFNMVYMPASWVMRIMTWLDMVDTSLSGFKLKDCVMYASAVVLLILMTARTVYDDGARRVW
TLMNVLTLVYKVYYGNALDQAISMWALIISVTSNYSGVVTTVMFLARGIVFMCVEYCPIFFITGNTLQCI
MLVYCFLGYFCTCYFGLFCLLNRYFRLTLGVYDYLVSTQEFRYMNSQGLLPPKNSIDAFKLNIKLLGVGG
KPCIKVATVQ

>NP_055935.4 Alstrom syndrome protein 1 isoform 1 [Homo sapiens]
MEPEDLPWPGELEEEEEEEEEEEEEEEEEAAAAAAANVDDVVVVEEVEEEAGRELDSDSHYGPQHLESID
DEEDEEAKAWLQAHPGRILPPLSPPQHRYSEGERTSLEKIVPLTCHVWQQIVYQGNSRTQISDTNVVCLE
TTAQRGSGDDQKTESWHCLPQEMDSSQTLDTSQTRFNVRTEDTEVTDFPSLEEGILTQSENQVKEPNRDL
FCSPLLVIQDSFASPDLPLLTCLTQDQEFAPDSLFHQSELSFAPLRGIPDKSEDTEWSSRPSEVSEALFQ
ATAEVASDLA

In [55]:
handle = Entrez.efetch(db="protein", id=swiss_prot_ids, rettype="fasta")

#save sequences to fasta file
fasta_file = "negative_proteins.fasta"
with open(fasta_file, "w") as fo:
    fo.write(handle.read())

#check that everything worked
!cat negative_proteins.fasta

>sp|Q13084.4|RM28_HUMAN RecName: Full=39S ribosomal protein L28, mitochondrial; Short=L28mt; Short=MRP-L28; AltName: Full=Melanoma antigen p15; AltName: Full=Melanoma-associated antigen recognized by T-lymphocytes; AltName: Full=Mitochondrial large ribosomal subunit protein bL28m; Flags: Precursor
MPLHKYPVWLWKRLQLREGICSRLPGHYLRSLEEERTPTPVHYRPHGAKFKINPKNGQRERVEDVPIPIY
FPPESQRGLWGGEGWILGQIYANNDKLSKRLKKVWKPQLFEREFYSEILDKKFTVTVTMRTLDLIDEAYG
LDFYILKTPKEDLCSKFGMDLKRGMLLRLARQDPQLHPEDPERRAAIYDKYKEFAIPEEEAEWVGLTLEE
AIEKQRLLEEKDPVPLFKIYVAELIQQLQQQALSEPAVVQKRASGQ

>sp|Q9Y5S9.1|RBM8A_HUMAN RecName: Full=RNA-binding protein 8A; AltName: Full=Binder of OVCA1-1; Short=BOV-1; AltName: Full=RNA-binding motif protein 8A; AltName: Full=RNA-binding protein Y14; AltName: Full=Ribonucleoprotein RBM8A
MADVLDLHEAGGEDFAMDEDGDESIHKLKEKAKKRKGRGFGSEEGSRARMREDYDSVEQDGDEPGPQRSV
EGWILFVTGVHEEATEEDIHDKFAEYGEIKNIHLNLDRRTGYLKGYTLVEYETYKEAQAAMEGLNGQDLM
GQPISVDWCFVRGPPKGKRRGGRRRSRSPDRRRR

>sp|Q9H8H0.1|NOL11_HUMAN RecName

In [56]:
# Analysis of the fasta file
from Bio import SeqIO
record = SeqIO.parse(open(fasta_file, 'r'),"fasta")

updated_seqs = []
for seq in record:
    seq_descr = seq.description
    full_name_section = seq_descr.split(";")[0]
    full_name = full_name_section.split("Full=")[1]
    seq.description = full_name
    print("ID: {}, Length: {}, Description: {}".format(seq.id, len(seq), seq.description))
    updated_seqs.append(seq)

ID: sp|Q13084.4|RM28_HUMAN, Length: 256, Description: 39S ribosomal protein L28, mitochondrial
ID: sp|Q9Y5S9.1|RBM8A_HUMAN, Length: 174, Description: RNA-binding protein 8A
ID: sp|Q9H8H0.1|NOL11_HUMAN, Length: 719, Description: Nucleolar protein 11
ID: sp|P49756.3|RBM25_HUMAN, Length: 843, Description: RNA-binding protein 25
ID: sp|P0C7T5.1|ATX1L_HUMAN, Length: 689, Description: Ataxin-1-like
ID: sp|Q8TDN6.2|BRX1_HUMAN, Length: 353, Description: Ribosome biogenesis protein BRX1 homolog
ID: sp|Q6NZY4.2|ZCHC8_HUMAN, Length: 707, Description: Zinc finger CCHC domain-containing protein 8
ID: sp|Q6NXG1.2|ESRP1_HUMAN, Length: 681, Description: Epithelial splicing regulatory protein 1
ID: sp|Q9P275.4|UBP36_HUMAN, Length: 1123, Description: Ubiquitin carboxyl-terminal hydrolase 36
ID: sp|P22314.3|UBA1_HUMAN, Length: 1058, Description: Ubiquitin-like modifier-activating enzyme 1
ID: sp|O95628.3|CNOT4_HUMAN, Length: 575, Description: CCR4-NOT transcription complex subunit 4
ID: sp|P51114.3|FXR1_

In [58]:
with open("negative_proteins.fasta", "w") as output_handle:
    SeqIO.write(updated_seqs, output_handle, "fasta")

## Next Steps
* Figuring out how to train a [transformer model](https://github.com/facebookresearch/esm) or [protein-based BERT model](https://github.com/nadavbra/protein_bert).
* But first, save the FASTA sequences to files!
* It is best not to run these lines again since they'll change the analysis. Consider this the final selection of negative training data for the model.

In [59]:
#!cp negative_proteins.fasta /content/drive/MyDrive/ecbm-e4060-covid-interactions/DATA/ML_Extension/negative_proteins.fasta

In [33]:
#!cp positive_proteins.fasta /content/drive/MyDrive/ecbm-e4060-covid-interactions/DATA/ML_Extension/positive_proteins.fasta